internal/manifest/version.go

```diff
@@ -256,31 +256,39 @@ func (v *Version) Next() *Version {
 // may touch). If level is zero then that assumption cannot be made, and the
 // [start, end] range is expanded to the union of those matching ranges so far
 // and the computation is repeated until [start, end] stabilizes.
+// The returned files are a subsequence of the input files, i.e., the ordering
+// is not changed.
 func (v *Version) Overlaps(
 	level int, cmp Compare, start, end []byte,
 ) (ret []FileMetadata) {
 	if level == 0 {
-		// The sstables in level 0 can overlap with each other. As soon as we find
-		// one sstable that overlaps with our target range, we need to expand the
-		// range and find all sstables that overlap with the expanded range.
-	loop:
+		// Indices that have been selected as overlapping.
+		selectedIndices := make([]bool, len(v.Files[level]))
+		numSelected := 0
+		restart := false
 		for {
-			for _, meta := range v.Files[level] {
+			for i, selected := range selectedIndices {
+				if selected { continue }
+				meta := &v.Files[level][i]
 				smallest := meta.Smallest.UserKey
 				largest := meta.Largest.UserKey
 				if cmp(largest, start) < 0 {
-					// meta is completely before the specified range; skip it.
+					// file is completely before the specified range; skip it.
 					continue
 				}
 				if cmp(smallest, end) > 0 {
-					// meta is completely after the specified range; skip it.
+					// file is completely after the specified range; skip it.
 					continue
 				}
-				ret = append(ret, meta)
-
-				// If level == 0, check if the newly added fileMetadata has
-				// expanded the range. If so, restart the search.
-				restart := false
+				// Overlaps
```

Comment:
Nit: we try to follow the Go convention of terminating comments with a period (except if they are the same line as code). I can't say we're perfect at this, but we try.

Reply:
done